## Importing necessary libraries

In [2]:
import pandas as pd

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score,r2_score


In [3]:
Zoo_data = pd.read_csv("Zoo.csv")
Zoo_data

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [5]:
Zoo_data.columns

Index(['animal name', 'hair', 'feathers', 'eggs', 'milk', 'airborne',
       'aquatic', 'predator', 'toothed', 'backbone', 'breathes', 'venomous',
       'fins', 'legs', 'tail', 'domestic', 'catsize', 'type'],
      dtype='object')

## Problem Statement : Implement a KNN model to classify the animals in to categorie

## Input :  animal name - hair - feathers - eggs - milk - airborne - aquatic - predator - toothed backbone - breathes - venomous - fins - legs - tail - domestic - catsize 

## Output : type

## Data understanding

In [6]:
Zoo_data.shape

(101, 18)

In [7]:
Zoo_data.describe()

hair    feathers        eggs        milk    airborne     aquatic  \
count  101.000000  101.000000  101.000000  101.000000  101.000000  101.000000   
mean     0.425743    0.198020    0.584158    0.405941    0.237624    0.356436   
std      0.496921    0.400495    0.495325    0.493522    0.427750    0.481335   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    1.000000    0.000000    0.000000    0.000000   
75%      1.000000    0.000000    1.000000    1.000000    0.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         predator     toothed    backbone    breathes    venomous        fins  \
count  101.000000  101.000000  101.000000  101.000000  101.000000  101.000000   
mean     0.554455    0.603960    0.821782    0.792079    0.079208    0.168317   
std      0.499505    0.491512    0.384605    0.407844    0.271410    0.376013   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    1.000000    1.000000    0.000000    0.000000   
50%      1.000000    1.000000    1.000000    1.000000    0.000000    0.000000   
75%      1.000000    1.000000    1.000000    1.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

             legs        tail    domestic     catsize        type  
count  101.000000  101.000000  101.000000  101.000000  101.000000  
mean     2.841584    0.742574    0.128713    0.435644    2.831683  
std      2.033385    0.439397    0.336552    0.498314    2.102709  
min      0.000000    0.000000    0.000000    0.000000    1.000000  
25%      2.000000    0.000000    0.000000    0.000000    1.000000  
50%      4.000000    1.000000    0.000000    0.000000    2.000000  
75%      4.000000    1.000000    0.000000    1.000000    4.000000  
max      8.000000    1.000000    1.000000    1.000000    7.000000

In [8]:
Zoo_data.dtypes

animal name    object
hair            int64
feathers        int64
eggs            int64
milk            int64
airborne        int64
aquatic         int64
predator        int64
toothed         int64
backbone        int64
breathes        int64
venomous        int64
fins            int64
legs            int64
tail            int64
domestic        int64
catsize         int64
type            int64
dtype: object

In [9]:
Zoo_data.isnull().sum()

animal name    0
hair           0
feathers       0
eggs           0
milk           0
airborne       0
aquatic        0
predator       0
toothed        0
backbone       0
breathes       0
venomous       0
fins           0
legs           0
tail           0
domestic       0
catsize        0
type           0
dtype: int64

## Data Preocessing 

## There is No null values so data cleaning not required

## Model Building 

In [4]:
Zoo_Label = LabelEncoder()
Zoo_data['animal name'] = Zoo_Label.fit_transform(Zoo_data['animal name'])

In [5]:
Zoo_data

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0              0     1         0     0     1         0        0         1   
1              1     1         0     0     1         0        0         0   
2              2     0         0     1     0         0        1         1   
3              3     1         0     0     1         0        0         1   
4              4     1         0     0     1         0        0         1   
..           ...   ...       ...   ...   ...       ...      ...       ...   
96            95     1         0     0     1         0        0         0   
97            96     1         0     1     0         1        0         0   
98            97     1         0     0     1         0        0         1   
99            98     0         0     1     0         0        0         0   
100           99     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [6]:
X = Zoo_data.drop(['type'],axis=1)
y = Zoo_data['type']

In [7]:
X.shape,y.shape

((101, 17), (101,))

In [8]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state= 15,stratify=y)
X_train.shape,y_train.shape,X_test.shape,y_test.shape

((80, 17), (80,), (21, 17), (21,))

In [9]:
Zoo_knn_model = KNeighborsClassifier()
Zoo_knn_model.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

## Model Testing

In [10]:
y_pred = Zoo_knn_model.predict(X_test)
y_pred

array([2, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 7, 1, 1, 1, 1, 1, 4, 2, 1, 1],
      dtype=int64)

## Model Evaulation 

In [11]:
print("Accuracy score   : ", round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   :  0.4286
R2 Score         : -0.8504


## Lets Choose best K value , because by defalut it taken K=5

In [18]:
neighbours = list(range(1,50,2))
cv_scores = []

for i in neighbours:
    Zoo_knn_model = KNeighborsClassifier(n_neighbors = i)
    cv_score = cross_val_score(estimator = Zoo_knn_model,X=X,y=y,cv=10)
    cv_scores.append(cv_score.mean())

In [13]:
import warnings
warnings.filterwarnings('ignore')

In [14]:
cv_scores

[0.4845454545454545,
 0.4445454545454545,
 0.4154545454545454,
 0.3854545454545455,
 0.3354545454545455,
 0.35545454545454547,
 0.3654545454545454,
 0.38545454545454544,
 0.37545454545454543,
 0.3854545454545455,
 0.3954545454545455,
 0.3954545454545454,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546]

In [15]:
neighbours[cv_scores.index(max(cv_scores))]

1

## As per above line to get better result K should be K=1

In [17]:
Zoo_knn_model = KNeighborsClassifier(n_neighbors=1)
Zoo_knn_model.fit(X_train,y_train)
y_pred = Zoo_knn_model.predict(X_test)
print("Accuracy score   :", round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   : 0.6667
R2 Score         : 0.333


## If we take K=5 then Accuracy is 0.4286
## If we take K=1 then Accuracy is 0.6667

## Model Depolyment

In [57]:
from pickle import dump

In [58]:
dump(Zoo_knn_model,open("Zoo_knn_model.pkl",'wb'))

In [59]:
from pickle import load

In [60]:
Zoo_loaded = load(open("Zoo_knn_model.pkl",'rb'))

In [61]:
Zoo_loaded.predict(X_test)

array([2, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 7, 1, 1, 1, 1, 1, 4, 2, 1, 1],
      dtype=int64)